src/engines/seq/euclid/euclid.cpp

```diff
@@ -28,33 +28,38 @@ namespace otto::engines {
   audio::ProcessData<0> Euclid::process(audio::ProcessData<0> data)
   {
     auto& current = current_channel();
+
+    // Will record up to 6 notes (firsts pressed within a chord)
+    // Possible to redo until explicitely exiting recording mode
+    // Will reset previous notes upon new entry
     if (recording) {
       for (auto& event : data.midi) {
         util::match(event,
                     [&](midi::NoteOnEvent& ev) {
-                      if (recording.value() == current.notes &&
-                          util::all_of(recording.value(), [](char note) { return note < 0; })) {
-                        util::fill(recording.value(), -1);
+                      // new event on an empty recording buffer: reset notes
+                      if (util::all_of(recording.value(), [](char note) { return note < 0; })) {
                         util::fill(current.notes, -1);
                       }
+                      // adding first new notes to buffer
                       for (auto& note : recording.value()) {
-                        if (note >= 0) continue;
+                        if (note >= 0 && note != ev.key) {
+                          continue;
```

Comment:
wont work. It will take an empty slot before one with the note already in it

Reply:
Upon release of a note the slot is set back to -1, this is only a problem if the same note is held twice at the same time... and indeed I could reproduce that by having two MIDI inputs. :sweat: